### PR TITLE
RFC: postbox

### PR DESCRIPTION
**Pull request by pontaoski**
After much discussion, this is the mostly finalised design for the postbox mechanic.

It consists of a pulling+pushing system that delivers events in a guaranteed order.

### CODE REVIEW

File: sync/v1/sync.proto [hunk 1]
```diff
@@ -1,25 +1,106 @@
...
-  string host = 2;
-}
+message Event {
+  message UserRemovedFromGuild {
+    uint64 user_id = 1;
+    uint64 guild_id = 2;
+  }
 
-message PostBoxEvent { google.protobuf.Any event = 1; }
+  oneof kind {
```
Comment (yusdacra) [inline sync/v1/sync.proto:15]: Don't we need more event kinds here? Like `UserAddedToGuild`?

File: sync/v1/sync.proto [hunk 2]
```diff
@@ -1,25 +1,106 @@
...
+// Self is the server name of the server initiating the transaction. For Pull,
+// this tells the server being connected to which homeservers' events it should send.
+// For Push, this tells the server being connected to which homeservers' events it is
+// receiving.
+//
+// Time is the UNIX time in seconds of when the request is started. Servers should reject
+// JWTs with a time too far from the current time, at their discretion. A recommended
+// variance is 1 minute.
+//
+// Rand is a random string. Servers must reject requests that re-use a Rand after
```
Comment (yusdacra) [inline sync/v1/sync.proto:67]: What are the conditions on `Rand` usage? Should servers ever clear usage of `Rand`?

**Review by yusdacra — CHANGES_REQUESTED** [review]

File: sync/v1/sync.proto (hunk 2 above)
Comment (pontaoski) [reply sync/v1/sync.proto]: Rand is kinda pointless for preventing replay attacks w/ the UNIX timestamp

**Review by yusdacra — APPROVED** [review]

File: sync/v1/sync.proto [hunk 3]
```diff
@@ -1,25 +1,107 @@
...
+// Self string
+// Time uint53
+// ```
+//
+// Self is the server name of the server initiating the transaction. For Pull,
+// this tells the server being connected to which homeservers' events it should send.
+// For Push, this tells the server being connected to which homeservers' events it is
+// receiving.
+//
+// Time is the UNIX time in seconds of when the request is started. Servers should reject
```
Comment (Bluskript) [inline sync/v1/sync.proto:67]: should probably clarify that this should not be relative to your time, rather UTC

**Review by Bluskript — APPROVED** [review]

**Pull request merged**